pkg/virt-dhcp/dnsmasq.go

```diff
@@ -0,0 +1,323 @@
+package virtdhcp
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"os"
+	"os/exec"
+	"sort"
+	"strconv"
+	"strings"
+	"syscall"
+	"text/template"
+
+	"kubevirt.io/kubevirt/pkg/log"
+)
+
+const (
+	DNSmasqExec = "/usr/sbin/dnsmasq"
+	DHCPPidFile = "/tmp/dhcp.pid"
+)
+
+var DHCPLeaseFile = "/var/run/kubevirt/dhcp.leases"
+var DHCPHostsFile = "/var/run/kubevirt/dhcp.hosts"
+
+type hostsDetails struct {
+	Mac   string
+	Lease int
+}
+
+type DNSmasqInstance struct {
+	hosts    map[string]hostsDetails
+	ipRange  [][]byte
+	confpath string
+	args     []string
+	cmd      *exec.Cmd
+	monitor  Monitor
+}
+
+func NewDNSmasq() *DNSmasqInstance {
+	dnsMasq := new(DNSmasqInstance)
+	dnsMasq.hosts = make(map[string]hostsDetails)
+	dnsMasq.ipRange = make([][]byte, 2)
+	dnsMasq.args = []string{"-k", "-d", "--strict-order", "--bind-dynamic", "--no-resolv", "--dhcp-no-override", "--dhcp-authoritative",
+		"--except-interface=lo", "--dhcp-hostsfile=" + DHCPHostsFile, "--dhcp-leasefile=" + DHCPLeaseFile,
+		"--pid-file=" + DHCPPidFile}
+	dnsMasq.monitor = NewMonitor()
+	OS = &OSHandler{}
+	return dnsMasq
+}
+
+// implement `Interface` in sort package.
+type sortByteArrays [][]byte
+
+func (b sortByteArrays) Len() int {
+	return len(b)
+}
+
+func (b sortByteArrays) Less(i, j int) bool {
+	// bytes package already implements Comparable for []byte.
+	switch bytes.Compare(b[i], b[j]) {
+	case -1:
+		return true
+	case 0, 1:
+		return false
+	default:
+		return false
+	}
+}
+
+func (b sortByteArrays) Swap(i, j int) {
+	b[j], b[i] = b[i], b[j]
+}
+
+type OSOps interface {
+	IsFileExist(path string) (bool, error)
+	getPidFromFile(file string) (int, error)
+	checkProcessExist(pid int) (bool, *os.Process)
+	killProcessIfExist(pid int) error
+	RecreateHostsFile() (*os.File, error)
+	executeCommand(string, []string) *exec.Cmd
+	readFromFile(file string) ([]byte, error)
+	IsDNSMasqPidRunning(pid int) bool
+}
+
+type OSHandler struct{}
+
+var OS OSOps
+
+func (o *OSHandler) IsFileExist(path string) (bool, error) {
+	_, err := os.Stat(path)
+	exists := false
+
+	if err == nil {
+		exists = true
+	} else if os.IsNotExist(err) {
+		err = nil
+	}
+	return exists, err
+}
+
+func (o *OSHandler) readFromFile(file string) ([]byte, error) {
+	content, err := ioutil.ReadFile(file)
+	if err != nil {
+		log.Log.Reason(err).Errorf("failed to open file: %s", file)
+		return nil, err
+	}
+	return content, nil
+}
+
+func (o *OSHandler) getPidFromFile(file string) (int, error) {
+	content, err := OS.readFromFile(file)
+	if err != nil {
+		return -1, err
+	}
+	lines := strings.Split(string(content), "\n")
+	pid, _ := strconv.Atoi(lines[0])
+	return pid, nil
+}
+
+func (o *OSHandler) killProcessIfExist(pid int) error {
+	procExist, proc := OS.checkProcessExist(pid)
+	if procExist {
```

Comment:
I think all the process detection in the normal start/stop flow is an absolute overkill. You have to make sure that you kill a left over dnsmasq server when this binary restarts (including the check mentioned by david  below to make sure that you don't kill the wrong process). Then you are in full control over the process you start. You never have to search it, just use the functinality of *Process directly.


`*Process` has a `Wait()` method. So after you started it, you can just spawn a go-routine and wait for it to terminate. Just protect the expected restarts with a lock, to avoid the "detection" of a crash. Further you can use `Exited()` for an explicit check. That all is now possible, since the dhcp server will then always be a child of the main process.

It will make the whole application much smaller and easier.

Reply:
Just so that it is mentioned: Since the dhcp server gets restarted if new mappings are added, the new process after the restart would be a child the main thread too.

Reply:
pre-existing process cleanup is only performed once at startup now.